packages/hardhat-etherscan/src/index.ts

```diff
@@ -272,12 +288,20 @@ Possible causes are:
   // TODO: Add known edge cases here.
   // E.g:
   // - "Unable to locate ContractCode at <address>"
-  // - Address of library used in constructor is wrong
-  throw new NomicLabsHardhatPluginError(
-    pluginName,
-    `The contract verification failed.
-Reason: ${verificationStatus.message}`
-  );
+  let errorMessage = `The contract verification failed.
+Reason: ${verificationStatus.message}`;
+  if (contractInformation.undetectableLibraries.length > 0) {
```

Comment:
Ah, yes; this is necessary to add it as a possible cause. Basically, if your contract uses libraries whose addresses are undetectable, we have no way of knowing whether the addresses provided by the user are okay until Etherscan verifies the contract or not. If the contract is verified, they are correct. But if the verification fails, the API itself isn't very informative about the cause of the failure. That's why I think it is necessary to include this reason in the error message conditionally. There is no corresponding check in `getLibraryLinks` for this since we have to assume these kinds of links correct.